### DISCUSSION TITLE
Problems with `flymake-no-changes-timeout` and `eglot-send-changes-idle-time`

### DISCUSSION THREAD

**Discussion by roife** [joaotavora/eglot#1083] [Q&A] [answered]
I use `flymake` as my syntax checker with `eglot`. And I added `(setq flymake-no-changes-timeout nil)` to my config so that `flymake` runs checker only on saving. Meanwhile, I also added `(setq eglot-send-changes-idle-time 2)` to config to reduce the performance pressure.

But I found these two variables may cause problems. After I finish typing, I have to wait for **2 seconds** (as the value of `eglot-send-changes-idle-time`) before saving to get correct diagnostics. If I save the file immediately after I stop typing, the diagnostics I get is produced from the code 2 seconds ago, which is improper.

The problem can be fixed by removing `(setq flymake-no-changes-timeout nil)` from my config or using `(setq eglot-send-changes-idle-time 0)`. But I wonder if there is any other solutions, such as making `eglot` `send changes` immediately when saving files.

**Comment by joaotavora** [comment, marked as answer]
roife ***@***.***> writes:

> The problem can be fixed by removing (setq flymake-no-changes-timeout
> nil) from my config or using (setq eglot-send-changes-idle-time
> 0). But I wonder if there is any other solutions, such as making eglot
> send changes immediately when saving files.

I think it does something equivalent already, but that wouldn't solve
you problem anyway.

When you save the file, Eglot tells the server that it did so, and
presumably the server can just consult the file's contents in the
filesystem to determine the diagnostics.  Perhaps it does this, but if
you have flymake-no-changes-timeout nil (and run the checker only on
saving), this is probably still not workable as the checker will have no
information to provide to you.

New LSP versions support so-called "pull diagnostics" that may ameliorate
the problem.  But not all servers support this necessarily.

My recomendation is to keep flymake-no-changes-timeout at its default
setting.  This is because, if your goal is to fix potentital performance
problems, then having Flymake ask Eglot for more diagnostics more often
is probably not a cause of performance problems.  So keep
flymake-no-changes-timeout somewhere non-nil and tweak
eglot-send-changes-idle-time as you see fit.

**Comment by tshu-w** [comment]
This [hack](https://github.com/joaotavora/eglot/issues/1296#issuecomment-1727978307) works for me.
